internal/pkg/service/stream/storage/level/local/reader/reader.go

```diff
@@ -8,55 +8,181 @@ package reader
 import (
 	"context"
 	"io"
-	"os"
+	"sync"
 
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/ctxattr"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/compression"
+	compressionReader "github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/compression/reader"
+	compressionWriter "github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/compression/writer"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/events"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/reader/readchain"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
 )
 
 type Reader interface {
-	io.ReadCloser
-
-	CloseCtx(ctx context.Context) error
-
 	SliceKey() model.SliceKey
 
-	// UnwrapFile returns underlying file, if it is the only element in the chain.
-	//
-	// This is preferred way, it enables internal Go optimization and "zero CPU copy" to be used,
-	// read more about "sendfile" syscall for details.
-	//
-	// The Close methods should always be called on the Chain, not directly on the File,
-	// because Chain may contain multiple closers, even if there is only one reader.
-	UnwrapFile() (f *os.File, ok bool)
-
-	// DirPath is absolute path to the slice directory. It contains slice file and optionally an auxiliary files.
-	DirPath() string
-	// FilePath is absolute path to the slice file.
-	FilePath() string
+	WriteTo(w io.Writer) (n int64, err error)
+	// Close all readers in the chain.
+	Close(ctx context.Context) error
+	// Events provides listening to the reader lifecycle.
+	Events() *events.Events[Reader]
 }
 
 type readChain = readchain.Chain
 
 type reader struct {
-	*readChain
-	sliceKey model.SliceKey
-	dirPath  string
-	filePath string
+	logger log.Logger
+	chain  *readChain
+	slice  *model.Slice
+	events *events.Events[Reader]
+
+	// closed blocks new reads
+	closed chan struct{}
+	// readWg waits for in-progress reads before Close
+	readWg *sync.WaitGroup
+}
+
+func New(
```

Comment:
- Reader is simpler than Writer, there is no need to deal with Flush/Sync operations.
- After reading all data, it is necessary to `Close` all readers in the chain to free memory.
- The order of the `Close` method calls practically does not matter.

-----

I added:
- `events` - it can be useful in the future, it has no use-case now
   - Now it match the writer pkg.
 - `closed` channal to prevent reads from a closed reader.
    - It's just an extra "fuse", if some other part of the app is written incorrectly.
    - Now it match the writer pkg.
 - `readWg` - wait for all in-flight reads before `close`.
    - It's just an extra "fuse", if some other part of the app is written incorrectly.
    - Now it match the writer pkg.